linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala

```diff
@@ -563,9 +563,19 @@ private class FunctionEmitter private (
   def genTreeToAny(tree: Tree): Unit =
     genTree(tree, if (tree.tpe.isNullable) AnyType else AnyNotNullType)
 
-  def genTree(tree: Tree, expectedType: Type): Unit = {
+  def genTree(tree: Tree, expectedType: Type, allowCast: Boolean = false): Unit = {
+    /* Note that while `VoidType` is not a valid target type for a `Cast`
+     * transient, it *is* a valid value for the `castTo` parameter of
+     * `genJS*` methods. Therefore, we can always use the `expectedType` as
+     * argument for `castTo`.
+     */
```

Comment:
Also add a comment that `allowCast` is best effort, so it is OK to cast even if it is not allowed? (IIUC we are actually doing that with `castTo`, but the benefits outweigh the downsides of doing so, or at least so I believe).